test/eventinge2erekt/features/encryption.go

```diff
@@ -0,0 +1,202 @@
+package features
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/injection/clients/dynamicclient"
+	"knative.dev/pkg/logging"
+	"knative.dev/reconciler-test/pkg/environment"
+	"knative.dev/reconciler-test/pkg/feature"
+	"knative.dev/reconciler-test/pkg/resources/knativeservice"
+)
+
+func VerifyEncryptedTrafficToActivatorToApp(refs []corev1.ObjectReference, since time.Time) *feature.Feature {
+	// Just a small sleep to let the logs be written.
+	time.Sleep(10 * time.Second)
+
+	f := feature.NewFeature()
+
+	f.Stable("path to activator to app").
+		Must("has encrypted traffic", func(ctx context.Context, t feature.T) {
+			env := environment.FromContext(ctx)
+
+			var (
+				ksvcName string
+				numKsvc  int
+			)
+			for _, ref := range refs {
+				if ref.GroupVersionKind().GroupVersion() == knativeservice.GVR().GroupVersion() {
+					if numKsvc != 0 {
+						t.Fatalf("Found more than one Knative Service: %s, %s", ksvcName, ref.Name)
+					}
+					ksvcName = ref.Name
+					numKsvc++
+				}
+			}
+
+			ksvc, err := dynamicclient.Get(ctx).Resource(knativeservice.GVR()).Namespace(env.Namespace()).
+				Get(ctx, ksvcName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("Unable to get ksvc %s: %v", ksvcName, err)
+			}
+
+			address, _, _ := unstructured.NestedString(ksvc.Object, "status", "address", "url")
+			privateURL, err := url.Parse(address)
+			if err != nil {
+				t.Fatalf("Unable to parse URL %s: %v", address, err)
+			}
+
+			// source -> activator
+			// When running within Mesh a mesh-specific VirtualService is used which
+			// gets istio-ingressgateway out of the path.
+			err = VerifyPodLogsEncryptedRequestToHost(ctx,
+				"knative-serving",
+				metav1.ListOptions{LabelSelector: "app=activator"},
+				&corev1.PodLogOptions{Container: "istio-proxy", SinceTime: &metav1.Time{Time: since}},
+				func(m map[string]interface{}) bool {
+					return getMapValueAsString(m, "path") == "/" &&
+						getMapValueAsString(m, "authority") == privateURL.Host
+				})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// activator -> application
+			err = VerifyPodLogsEncryptedRequestToHost(ctx,
+				env.Namespace(),
+				metav1.ListOptions{LabelSelector: "serving.knative.dev/service=" + ksvcName},
+				&corev1.PodLogOptions{Container: "istio-proxy", SinceTime: &metav1.Time{Time: since}},
+				func(m map[string]interface{}) bool {
+					return getMapValueAsString(m, "path") == "/" &&
+						strings.HasPrefix(getMapValueAsString(m, "upstream_cluster"), "inbound|80")
+				})
+			if err != nil {
+				t.Fatal(err)
+			}
```

Comment:
I'd make the different paths, 2 isolated `StepFn` and the test becomes:
```
f.Stable("path to activator to app").
  Must("have encrypted traffic to activator", VerifyPodLogsEncryptedRequestToHost( ...)).
  Must("have encrypted traffic to app", VerifyPodLogsEncryptedRequestToHost( ...))

Reply:
2 advantages:
- it's more structured towards features and individual verification steps (ideally favoring reusing if we ever need to)
- steps might run in parallel

Reply:
bonus point, the test names tells what part actually failed